wagtail/models/__init__.py

```diff
@@ -2903,26 +2959,99 @@ class GroupPagePermission(models.Model):
         related_name="group_permissions",
         on_delete=models.CASCADE,
     )
+    permission = models.ForeignKey(
+        Permission,
+        verbose_name=_("permission"),
+        null=True,
+        blank=True,
+        on_delete=models.CASCADE,
+    )
     permission_type = models.CharField(
         verbose_name=_("permission type"),
+        null=True,
+        blank=True,
         max_length=20,
         choices=PAGE_PERMISSION_TYPE_CHOICES,
     )
 
+    objects = GroupPagePermissionManager()
+
     class Meta:
-        unique_together = ("group", "page", "permission_type")
+        constraints = [
+            models.CheckConstraint(
+                check=(
+                    models.Q(permission__isnull=False)
+                    | models.Q(permission_type__isnull=False)
+                ),
+                name="permission_or_permission_type_not_null",
+            ),
+            models.UniqueConstraint(
+                fields=("group", "page", "permission"),
+                name="unique_permission",
+            ),
+            models.UniqueConstraint(
+                fields=("group", "page", "permission_type"),
+                name="unique_permission_type",
+            ),
+        ]
         verbose_name = _("group page permission")
         verbose_name_plural = _("group page permissions")
 
+    @classmethod
+    def check(cls, **kwargs):
+        messages = super().check(**kwargs)
+        try:
+            outdated_objs = cls.objects._migrate_permission_type()
+        except OperationalError:
```

Comment:
On bakerydemo under Postgres with a clean database, this raises a `ProgrammingError` rather than an `OperationalError` for me. I suggest this is replaced with `django.db.DatabaseError`, which is the superclass of both of these.

Reply:
Ah, I suspected that but forgot to test, thanks! Rebased and fixed (along with other changes mentioned above) in 351ad3f.